subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractCrossTaskIncrementalJavaCompilationIntegrationTest.groovy

```diff
@@ -79,6 +236,171 @@ abstract class AbstractCrossTaskIncrementalJavaCompilationIntegrationTest extend
         impl.recompiledClasses('ImplB')
     }
 
+    def "changing an unused non-private constant doesn't cause any compilation"() {
+        source api: ["class A {}", "class B { final static int x = 1; }"], impl: ["class ImplA {}", "class ImplB {}"]
+        impl.snapshot { run language.compileTaskName }
+
+        when:
+        source api: ["class B { final static int x = 2; /* change */ void blah() { /* avoid flakiness by changing compiled file length*/ }  }"]
+        run "impl:${language.compileTaskName}"
+
+        then:
+        impl.noneRecompiled()
+    }
+
+    def "removing an unused non-private constant doesn't cause any compilation"() {
+        source api: ["class A {}", "class B { final static int x = 1; }"], impl: ["class ImplA {}", "class ImplB {}"]
+        impl.snapshot { run language.compileTaskName }
+
+        when:
+        source api: ["class B { /* change */ }"]
+        run "impl:${language.compileTaskName}"
+
+        then:
+        impl.noneRecompiled()
+    }
+
+    def "does not recompiles in case of conflicting changing constant values"() {
+        source api: ["class A { final static int x = 3; }", "class B { final static int x = 3; final static int y = -2; }"],
+            impl: ["class X { int foo() { return 3; }}", "class Y {int foo() { return -2; }}"]
+        impl.snapshot { run language.compileTaskName }
+
+        when:
+        source api: ["class B { final static int x = 3 ; final static int y = -3;  void blah() { /*  change irrelevant to constant value x */ } }"]
+        source api: ["class A { final static int x = 2 ; final static int y = -2;  void blah() { /*  change irrelevant to constant value y */ } }"]
+        run "impl:${language.compileTaskName}"
+
+        then:
+        impl.recompiledClasses()
+    }
+
+    @NotYetImplemented
+    // This would be possible but it's not yet implemented since we would have
+    // to track every constants as (symbol, value) pair and this is expensive, so there is no solution yet
+    def "ignores irrelevant changes to constant values in annotations"() {
```

Comment:
This would only work across task boundaries anyway. Inside a single compile task, we don't know which constant has changed (because we don't parse source files), so in that (common) case we'd recompile everything anyway.

Reply:
I see, thank you for clarifying this. I left this test for now, since it test cross task case and I believe this would be possible. What do you think?

Reply:
It would be, yes.